stdlib/public/core/ArrayCast.swift

```diff
@@ -16,181 +16,46 @@
+        src.downcast(toBufferWithDeferredTypeCheckOf: TargetElement.self))
     }
-    // All non-native buffers use deferred element typechecking
     return Array(_immutableCocoaArray: source._buffer._asCocoaArray())
-    
-  case (.reference, .explicit):
-    let result: [TargetElement]? = _arrayConditionalBridgeElements(source)
-    _precondition(result != nil, "array cannot be bridged from Objective-C")
-    return result!
-    
-  case (.value, .verbatim):
-    if source.isEmpty {
-      return Array()
-    }
-
-    var buf = _ContiguousArrayBuffer<TargetElement>(
-      uninitializedCount: source.count, minimumCapacity: 0)
-    
-    let _: Void = buf.withUnsafeMutableBufferPointer {
-      var p = $0.baseAddress!
-      for value in source {
-        let bridged: AnyObject? = _bridgeToObjectiveC(value)
-        _precondition(
-          bridged != nil, "array element cannot be bridged to Objective-C")
-        // FIXME: should be an unsafeDowncast.
-        p.initialize(with: unsafeBitCast(bridged!, to: TargetElement.self))
-        p += 1
-      }
-    }
-    return Array(_buffer: _ArrayBuffer(buf, shiftedToStartIndex: 0))
-    
-  case (.value, .explicit):
-    _sanityCheckFailure(
-      "Force-casting between Arrays of value types not prevented at compile-time"
-    )
   }
+  return _arrayConditionalCast(source)!
 }
 
-//===--- Conditional casts: [T] as? [U] -----------------------------------===//
-
-/// Implements the semantics of `x as? [TargetElement]` where `x` has type
-/// `[SourceElement]` and `TargetElement` is a verbatim-bridged trivial subtype of
-/// `SourceElement`.
-///
-/// Returns an Array<TargetElement> containing the same elements as a
-///
-/// O(1) if a's buffer elements are dynamically known to have type
-/// TargetElement or a type derived from TargetElement.  O(N)
-/// otherwise.
-internal func _arrayConditionalDownCastElements<SourceElement, TargetElement>(
-  _ a: Array<SourceElement>
-) -> [TargetElement]? {
-  _sanityCheck(_isBridgedVerbatimToObjectiveC(SourceElement.self))
-  _sanityCheck(_isBridgedVerbatimToObjectiveC(TargetElement.self))
-  
-  if _fastPath(!a.isEmpty) {
-    let native = a._buffer.requestNativeBuffer()
-    
-    if _fastPath(native != nil) {
-      if native!.storesOnlyElementsOfType(TargetElement.self) {
-        return Array(_buffer: a._buffer.cast(toBufferOf: TargetElement.self))
-      }
-      return nil
-    }
-    
-    // slow path: we store an NSArray
-    
-    // We can skip the check if TargetElement happens to be AnyObject
-    if !(AnyObject.self is TargetElement.Type) {
-      for element in a {
-        if !(element is TargetElement) {
-          return nil
-        }
-      }
-    }
-    return Array(_buffer: a._buffer.cast(toBufferOf: TargetElement.self))
+struct UnwrapFailure : Error {}
+extension Optional {
+  internal func unwrappedOrError() throws -> Wrapped {
+    if self == nil { throw UnwrapFailure() }
```

Comment:
Hey, D—

I don't see any unsafe APIs and if what you wrote compiles I don't know what it means